File: randomized_queue.py
import random

class RandomizedQueue:
    def __init__(self):
        self._array = [None]
        self._deleted_places = []
        self._tail = 0
        self._size = 0

    def is_empty(self):
        return self._size == 0
    
    @property
    def size(self):
        return self._size
    
    def enqueue(self, item):
        if item == None:
            raise Exception("Item is None")
        if len(self._deleted_places) > 0:
            self._array[self._deleted_places.pop()] = item
        if self._tail == len(self._array):
            self._array += [None] * len(self._array)
        self._array[self._tail] = item
        self._tail += 1
        self._size += 1

    def dequeue(self):
        if self.is_empty():
            raise Exception("RandomizedQueue is empty")
        index = random.randint(0, self._tail-1)
        while self._array[index] is None:
            index = random.randint(0, self._tail-1)
        self._deleted_places.append(index)
        item = self._array[index]
        self._array[index] = None
        self._size -= 1
        return item

    def sample(self):
        if self.is_empty():
            raise Exception("RandomizedQueue is empty")
        return self._array[random.randint(0, self._tail-1)]

    def __iter__(self):
        return self.RandomizedQueueIterator(self)
    
    class RandomizedQueueIterator:
        def __init__(self, queue):
            self._array = queue._array
            self._indices = [i for i, item in enumerate(self._array) if item is not None]
            self._size = queue._size
        def __next__(self):
            if not self._indices:
                raise StopIteration
            index = random.choice(self._indices)
            self._indices.remove(index)
            return self._array[index]
        def __iter__(self):
            return self

File: JavaFiles/week2/RandomizedQueue.java
import java.util.Iterator;
import java.util.NoSuchElementException;
import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.StdOut;

public class RandomizedQueue<Item> implements Iterable<Item> {
    private Item[] array;
    private Integer[] deletedPlaces;
    private int deletedTail;
    private int tail;
    private int size;

    
    public RandomizedQueue() {
        array = (Item[]) new Object[1];
        deletedPlaces = new Integer[1];
        deletedTail = 0;
        tail = 0;
        size = 0;
    }

    
    
    private void resize(int capacity) {
        Item[] temp = (Item[]) new Object[capacity];

        Integer[] deletedTemp = new Integer[capacity];
    
        System.arraycopy(array, 0, temp, 0, array.length);
        System.arraycopy(deletedPlaces, 0, deletedTemp, 0, deletedPlaces.length);
        array = temp;
        deletedPlaces = deletedTemp;
    }



    public boolean isEmpty() {
        return size == 0;
    }

    public int size() {
        return size;
    }

    public void enqueue(Item item) {
        if (item == null) {
            throw new IllegalArgumentException("Item is null");
        }
        if (deletedTail > 0) {
            array[deletedPlaces[--deletedTail]] = item;
        } else {     
            if (tail == array.length) { 
                resize(2 * array.length);
            }
            array[tail++] = item;
        }
        size++;
    }

    public Item dequeue() {
        if (isEmpty()) {
            throw new NoSuchElementException("RandomizedQueue is empty");
        }
        int index = StdRandom.uniform(tail);
        while (array[index] == null) {
            index = StdRandom.uniform(tail);
        }
        if (deletedTail == deletedPlaces.length) {
            resize(2 * array.length); // Note: You might want to adjust this logic
        }
        deletedPlaces[deletedTail++] = index;
        Item item = array[index];
        array[index] = null;
        size--;
        return item;
    }

    public Item sample() {
        if (isEmpty()) {
            throw new NoSuchElementException("RandomizedQueue is empty");
        }
        int index;
        do {
            index = StdRandom.uniform(tail);
        } while (array[index] == null);
        return array[index];
    }

    @Override
    public Iterator<Item> iterator() {
        return new RandomizedQueueIterator();
    }

    private class RandomizedQueueIterator implements Iterator<Item> {
        private final Item[] shuffledArray;
        private int currentIndex = 0;

        
        public RandomizedQueueIterator() {
            shuffledArray = (Item[]) new Object[size];
            int index = 0;
            for (Item item : array) {
                if (item != null) {
                    shuffledArray[index++] = item;
                }
            }
            StdRandom.shuffle(shuffledArray);
        }

        @Override
        public boolean hasNext() {
            return currentIndex < size;
        }

        @Override
        public Item next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            return shuffledArray[currentIndex++];
        }
    }

    public static void main(String[] args) {
        RandomizedQueue<Integer> queue = new RandomizedQueue<>();
        queue.enqueue(1);
        queue.enqueue(2);
        queue.enqueue(3);
        for (int item : queue) {
            StdOut.println(item);
        }
        System.out.println(queue.sample());
        System.out.println(queue.dequeue());
        System.out.println(queue.size());
        System.out.println(queue.isEmpty());
    }
}
